### PR TITLE
Adds agent API

### PR DESCRIPTION
**Pull request by bLopata**
Adds an simple, extensible API to allow for state sharing between client and agent, as well as arbitrary function registration and code execution 🚀

### CODE REVIEW

File: packages/ai/src/agent/index.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
+interface Agent {
+    llm: any;
+    systemPrompt?: string;
+    model: string;
+    contexts: Record<string, any>;
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
```

File: packages/ai/src/agent/index.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    functions: Record<string, Function>;
+    schemas: Array<any>;
+}
+
+const Agent = {
+    create: (llm: any, systemPrompt: string, model = 'gpt-3.5-turbo-0125'): Agent => {
+        const agent = {
+            llm,
+            model,
+            contexts: {},  // Key/value storage for state management
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
```

File: packages/ai/src/agent/index.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        return agent;
+    },
+
+    registerFunction: (agent: Agent, schema: any, fn: Function) => {
+        agent.functions[schema.function.name] = fn;
+        agent.schemas.push(schema);
+    },
+
+    // Running the agent with sequential function execution
+    run: async (agent: Agent, userMessage: string) => {
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
    run: async (agent: Agent, userMessage: string, options: {model?:string, temperature?:number, seed?: number}) => {
```

File: packages/ai/src/agent/index.ts [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        } try {
+            let lastResponse;
+            while (true) {
+                const response = await agent.llm.chat.completions.create({
+                    model: agent.model,
+                    messages: messages,
+                    tools: agent.schemas,
+                    temperature: 0.7,
+                    seed: 42,
+                });
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
                const response = await agent.llm.chat.completions.create({
                    model: (options && options.model) || 'gpt-4-turbo-preview',
                    messages: messages,
                    tools: agent.schemas,
                    temperature: options && options.temperature,
                    seed: options && options.seed,
                });
```

File: packages/ai/src/agent/index.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
+interface Agent {
+    llm: any;
+    systemPrompt?: string;
+    model: string;
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
```

File: packages/ai/src/agent/index.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    contexts: Record<string, any>;
+    functions: Record<string, Function>;
+    schemas: Array<any>;
+}
+
+const Agent = {
+    create: (llm: any, systemPrompt: string, model = 'gpt-3.5-turbo-0125'): Agent => {
+        const agent = {
+            llm,
+            model,
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
```

File: packages/ai/src/agent/index.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    llm: any;
+    systemPrompt?: string;
+    model: string;
+    contexts: Record<string, any>;
+    functions: Record<string, Function>;
+    schemas: Array<any>;
+}
+
+const Agent = {
+    create: (llm: any, systemPrompt: string, model = 'gpt-3.5-turbo-0125'): Agent => {
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
    create: (llm: any, systemPrompt: string): Agent => {
```

File: packages/ai/src/agent/index.ts [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+                        const functionResponse = await functionToCall(functionArgs);
+                        const functionMessage = {
+                            tool_call_id: toolCall.id,
+                            role: "tool",
+                            name: functionName,
+                            content: JSON.stringify(functionResponse),
+                        };
+
+                        messages.push(functionMessage); // Add function response
+                        lastResponse = await functionResponse; // Store the last response
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
                        lastResponse = functionResponse; // Store the last response
```

**Review by NateTheGreatt — COMMENTED** [review]
looks great! a few change requests:

File: packages/ai/src/agent/index.ts [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    },
+
+    registerFunction: (agent: Agent, schema: any, fn: Function) => {
+        agent.functions[schema.function.name] = fn;
+        agent.schemas.push(schema);
+    },
+
+    // Running the agent with sequential function execution
+    run: async (agent: Agent, userMessage: string, options: {model?:string, temperature?:number, seed?: number}) => {
+        let messages = [{ role: "user", content: userMessage }];
```
Comment (NateTheGreatt) [inline packages/ai/src/agent/index.ts]:
```suggestion
type RunOptions = {
    model?:string
    temperature?:number
    seed?: number
    messages?: any
}

    // Running the agent with sequential function execution
    run: async (agent: Agent, prompt: string, options: RunOptions) => {
        const message =  { role: "user", content: prompt }
        const messages = options.messages ? [...options.messages, message] : [message]
```

**Review by NateTheGreatt — COMMENTED** [review]
thought of something else that will be useful

File: packages/ai/src/agent/index.ts (hunk 9 above)
Comment (bLopata) [reply packages/ai/src/agent/index.ts]: Moar typing! 👍

Comment (NateTheGreatt) [reply packages/ai/src/agent/index.ts]: also can pass in `messages`

**Pull request merged**